### PR TITLE
tests: added another download client tests, use fff framework

### PR DESCRIPTION
**Pull request by chmielewskiandreas**
- Added another download client tests for socket recv/send timeout behavior.
- Refactored test to use fff framework.

### CODE REVIEW

**Comment by chmielewskiandreas** [comment]
any update here?

**Comment by gopiotr** [comment]
@rlubos @lemrey @simensrostad please perform review or point more suitable person to do this

File: tests/subsys/net/lib/download_client/CMakeLists.txt [hunk 1]
```diff
@@ -9,37 +9,25 @@ cmake_minimum_required(VERSION 3.20.0)
...
+target_include_directories(app PRIVATE ${ZEPHYR_BASE}/subsys/net/ip/)
+
+add_compile_definitions(CONFIG_COAP=1)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_LOG_LEVEL=4)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_HTTP_FRAG_SIZE=256)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE=32)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE=64)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_TCP_SOCK_TIMEO_MS=0)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_BUF_SIZE=0x40)
+add_compile_definitions(CONFIG_DOWNLOAD_CLIENT_STACK_SIZE=2048)
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/CMakeLists.txt:33]: Are those changes actually fixing something or is it just refacotring for the sake of refactoring? They're not mentioned at all in the commit message.

File: tests/subsys/net/lib/download_client/prj.conf [hunk 2]
```diff
@@ -1,21 +1,12 @@
-#
-# Copyright (c) 2022 Nordic Semiconductor ASA
-#
-# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
-#
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/prj.conf:5]: Please don't remove copyrights.

File: tests/subsys/net/lib/download_client/src/main.c [hunk 3]
```diff
@@ -1,19 +1,24 @@
 /*
- * Copyright (c) 2022 Nordic Semiconductor ASA
- *
- * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
- */
+* Copyright (c) 2022 grandcentrix GmbH
```
Comment (rlubos) [inline tests/subsys/net/lib/download_client/src/main.c:2]: And please don't replace copyrights. You can add your copyright just below the existing one.

**Review by rlubos — COMMENTED** [review]
Generally, it'd be good to split new tests and refactoring into separate commits, it's hard to review otherwise.

File: tests/subsys/net/lib/download_client/src/main.c [hunk 4]
```diff
@@ -95,128 +81,222 @@ static void dl_coap_start(struct download_client *client)
...
+					       ARRAY_SIZE(socket_sendto_ret_values));
 
-	dl_coap_start(&client);
+	download_start(&client);
 
-	zassert_ok(wait_for_event(DOWNLOAD_CLIENT_EVT_DONE, 10), "Download must have finished");
+	zassert_true(wait_for_event(DOWNLOAD_CLIENT_EVT_DONE), "Download must have finished");
+	zassert_equal(coap_get_recv_timeout_fake.call_count, 4, "");
+	zassert_equal(coap_request_send_fake.call_count, 4, "");
+	zassert_equal(coap_parse_fake.call_count, 3, "");
```
Comment (lemrey) [inline tests/subsys/net/lib/download_client/src/main.c:152]: You don't need to pass `""` to these macros, simply omit the parameter in the call.

**Review by lemrey — COMMENTED** [review]
If it adds tests, and they are run and are green I am happy with this.
I am not familiar with `FFF`, in fact I have never used it and I am also not using this library actively anymore. Perhaps @SeppoTakalo has some experience with `FFF` it and since he's using the `download_client` too he could give a quick review here?

**Comment by chmielewskiandreas** [comment]
> Generally, it'd be good to split new tests and refactoring into separate commits, it's hard to review otherwise.

Right, next time I will do that

**Comment by SeppoTakalo** [comment]
This looks to work OK on top of https://github.com/nrfconnect/sdk-nrf/pull/10442

**Comment by gopiotr** [comment]
@chmielewskiandreas do you plan to continue work with this PR? I ask because we need to accelerate work connected with migration to new Ztest API (it should be finished to the end of June). If you don't have a time to work with all issues connected with this PR, please split it into two PRs and make migration in separate one. If you don't have a time to do this, please let me know, and I will create separate PR with migration to new Ztest API for `tests/subsys/net/lib/download_client` test ;)

**Comment by chmielewskiandreas** [comment]
> @chmielewskiandreas do you plan to continue work with this PR? I ask because we need to accelerate work connected with migration to new Ztest API (it should be finished to the end of June). If you don't have a time to work with all issues connected with this PR, please split it into two PRs and make migration in separate one. If you don't have a time to do this, please let me know, and I will create separate PR with migration to new Ztest API for `tests/subsys/net/lib/download_client` test ;)

Sorry, I did not have any time recently to do something here. I have splitted the PR although the migration to fff test remains a larger change.

**Review by SeppoTakalo — APPROVED** [review]

**Comment by chmielewskiandreas** [comment]
> I have an error, when I try to run those tests on real hardware (tests pass on `native_posix`).
> 
> @chmielewskiandreas Should I do something extra before run those tests on my nrf91 DK?
> 
> Twister command:
> 
> ```
> ../zephyr/scripts/twister -vv --inline-logs -p nrf9160dk_nrf9160 -T tests/subsys/net/lib/download_client --device-testing --device-serial /dev/ttyACM0
> ```
> 
> Hardware output:
> 
> ```
> *** Booting Zephyr OS build v3.3.99-ncs1-rc2 ***
> Running TESTSUITE download_client
> ===================================================================
> START - test_download_abort_on_invalid_block_fail
> D: state = 1
> I: Downloading: no.file [0]
> E: CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE is 0
> W: Failed to resolve hostname 10.1.0.10 on IPv4
> Assertion failed at WEST_TOPDIR/nrf/tests/subsys/net/lib/download_client/src/main.c:230: download_client_test_download_abort_on_invalid_block_fail: coap_get_recv_timeout_fake.call_count not equal to 3
> FAIL - test_download_abort_on_invalid_block_fail in 0.124 seconds
> ===================================================================
> START - test_download_ignore_duplicate_block_ok
> D: state = 1
> I: Downloading: no.file [0]
> E: CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE is 0
> W: Failed to resolve hostname 10.1.0.10 on IPv4
> ```

can you try to add CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=2048 to prj.conf please and try again. Unfortunately I am not able to test this now

**Comment by chmielewskiandreas** [comment]
would you mind if I just remove support for nrf9160*?. There is no platform dependent test here, so it should be fine.

**Comment by gopiotr** [comment]
> would you mind if I just remove support for nrf9160*?. There is no platform dependent test here, so it should be fine.

Yes, eventually we can remove it, but maybe it is possible to fix this test for hardware. I made some experiments, and I observed, that there is probably a problem in `download_client_callback`. During `test_get` test, the `last_event` is not updated correctly, and that's why `test_get` "go down" on `wait_for_event(DOWNLOAD_CLIENT_EVT_DONE)`. 

I know, that it is not a best idea for debugging callbacks but just on a test purpose I added `printk` to `download_client_callback`  to observe what events are taken by this callback and surprisingly it fixed this test.
```c
static int download_client_callback(const struct download_client_evt *event)
{
	if (event == NULL) {
		return -EINVAL;
	}
	printk("TEST event callback %d\n", event->id);
	memcpy(&last_event, event, sizeof(struct download_client_evt));

	return 0;
}
```
Test output:
```
DEBUG   - DEVICE: ===================================================================
DEBUG   - DEVICE: START - test_get
DEBUG   - Could not find a matching testcase for net.lib.download_client.get
DEBUG   - DEVICE: D: state = 1
DEBUG   - DEVICE: I: Downloading: coap://192.168.1.2/large [0]
DEBUG   - DEVICE: D: Port not specified, using default: 5683
DEBUG   - DEVICE: D: family: 1, type: 2, proto: 17
DEBUG   - DEVICE: I: Connecting to coap://192.168.1.2/large
DEBUG   - DEVICE: D: fd 0, addrlen 8, fam IPv4, port 5683
DEBUG   - DEVICE: D: state = 2
DEBUG   - DEVICE: D: Receiving up to 64 bytes at 0x20004870...
DEBUG   - DEVICE: D: Read 25 bytes from socket
DEBUG   - DEVICE: I: Downloaded 25/75 bytes (33%)
DEBUG   - DEVICE: TEST event callback 0
DEBUG   - DEVICE: D: Receiving up to 64 bytes at 0x20004870...
DEBUG   - DEVICE: D: Read 25 bytes from socket
DEBUG   - DEVICE: I: Downloaded 50/75 bytes (66%)
DEBUG   - DEVICE: TEST event callback 0
DEBUG   - DEVICE: D: Receiving up to 64 bytes at 0x20004870...
DEBUG   - DEVICE: D: Read 25 bytes from socket
DEBUG   - DEVICE: I: Downloaded 75/75 bytes (100%)
DEBUG   - DEVICE: TEST event callback 0
DEBUG   - DEVICE: I: Download complete
DEBUG   - DEVICE: TEST event callback 2
DEBUG   - DEVICE: D: state = 4
DEBUG   - DEVICE: D: state = 0
DEBUG   - DEVICE: TEST event callback 3
DEBUG   - DEVICE: D: Connection closed
DEBUG   - DEVICE: PASS - test_get in 0.908 seconds
```
In this PR you changed a way, how `last_event` variable is updated in `download_client_callback` - now you use `memcpy` - is it safe in this context? 
Do you have any idea what can be a problem with updating `last_event` in callback?

**Comment by gopiotr** [comment]
@chmielewskiandreas any update/comment? If you don't have a time now to update/fix this PR in this week, then I can perform migration to new Ztest API in separate PR without introduction of fff mocking and new tests, there: https://github.com/nrfconnect/sdk-nrf/pull/11502.

I propose this solution, because we have to finish migration to new Ztest API to the end of this week, and thanks to this you could continue work connected with this PR in further future.

**Comment by chmielewskiandreas** [comment]
> @chmielewskiandreas any update/comment? If you don't have a time now to update/fix this PR in this week, then I can perform migration to new Ztest API in separate PR without introduction of fff mocking and new tests, there: #11502.
> 
> I propose this solution, because we have to finish migration to new Ztest API to the end of this week, and thanks to this you could continue work connected with this PR in further future.

yes, sure then I can continue working on this PR once I have time again. Thx :)

**Comment by gopiotr** [comment]
> yes, sure then I can continue working on this PR once I have time again. Thx :)

Okay, no problem :) my PR https://github.com/nrfconnect/sdk-nrf/pull/11502 was merged. If you will continue work in this PR, please rebase into the main.

**Pull request closed without merging**